### PR TITLE
The 'Jump to base' hotkey is useless if the Construction Yard is dest…

### PR DESCRIPTION
**Pull request by bryanwilbur**
…royed #3318

Now works like original Red Alert.  When there are no more Construction Yards the 'Jump to base' hotkey goes to the oldest undestroyed building.

### CODE REVIEW

File: OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs [hunk 1]
```diff
@@ -191,15 +191,27 @@ bool CycleBases()
...
+            if (!bases.Any())
+                return true;
 
 			var next = bases
 				.SkipWhile(b => !world.Selection.Actors.Contains(b))
 				.Skip(1)
 				.FirstOrDefault();
 
-			if (next == null)
+			if (next == null || noBase)
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs]: Why is this change?

File: OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs [hunk 2]
```diff
@@ -191,15 +191,27 @@ bool CycleBases()
 				.Select(b => b.Actor)
 				.ToList();
 
-			if (!bases.Any())
-				return true;
+            var noBase = false;
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs]: I think you don't even need this boolean at all. Merely the function below is enough.

**Comment by GraionDilach** [comment]
Also you apparently used wrong line endings in the commit. If you're using Visual Studio, you can either fix the file by using File->Advanced Save Options... and select Unix line endings there or use the Strip'em plugin: http://www.grebulon.com/software/stripem.php

File: OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs [hunk 3]
```diff
@@ -191,15 +191,28 @@ bool CycleBases()
...
+            if (!bases.Any())
+            {
+                bases = world.ActorsWithTrait<TargetableBuilding>()
+                .Where(a => a.Actor.Owner == world.LocalPlayer)
+                .Select(b => b.Actor)
+                .ToList();
+                noBase = true;
+            }
+
+            if (!bases.Any())
```
Comment (WolfGaming) [inline OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs]: You can move this if inside the other if so you don't check it twice if its false in both cases.

File: OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs (hunk 2 above)
Comment (WolfGaming) [inline OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs]: indeed the noBase variable shouldnt be needed.

**Comment by GraionDilach** [comment]
That and the snappet uses spaces for indenting instead of tabs. See https://msdn.microsoft.com/en-us/library/ms165330%28v=vs.90%29.aspx how to change that.

File: OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs [hunk 4]
```diff
@@ -191,15 +191,28 @@ bool CycleBases()
...
 				.ToList();
 
-			if (!bases.Any())
-				return true;
+            var noBase = false;
+
+            // If no base found the nondestroyed buildings are selected.
+            if (!bases.Any())
+            {
+                bases = world.ActorsWithTrait<TargetableBuilding>()
```
Comment (penev92) [inline OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs]: Is there a particular reason this uses `TargetableBuilding`?
That makes me uncomfortable for some reason. Can you use `Building`?

File: OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs [hunk 5]
```diff
@@ -191,15 +191,28 @@ bool CycleBases()
 				.Select(b => b.Actor)
 				.ToList();
 
-			if (!bases.Any())
-				return true;
+            var noBase = false;
+
+            // If no base found the nondestroyed buildings are selected.
```
Comment (penev92) [inline OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs]: "nondestroyed" -> "existing"
"If no base found" -> "If no base is found,"
aaand probably "the" -> "all" :D

File: OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs (hunk 4 above)
Comment (WolfGaming) [inline OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs]: Wouldn't it be weirder if this command allowed you to select buildings that aren't targetable?

Comment (penev92) [inline OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs]: We're talking about player-owned buildings here, so I don't see a case where you'd own something that is not targetable.

Comment (WolfGaming) [inline OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs]: We don't want to introduce edge cases here if a custom map assigns a non selectable building to a player.

**Comment by bryanwilbur** [comment]
Updated so boolean variable is removed and second if statement moved inside the first one.

File: OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs [hunk 6]
```diff
@@ -191,15 +191,24 @@ bool CycleBases()
...
+            }   
 
 			var next = bases
 				.SkipWhile(b => !world.Selection.Actors.Contains(b))
 				.Skip(1)
 				.FirstOrDefault();
 
-			if (next == null)
+            // If no construction yard exists then next is set to only the oldest existing building
+			if (next == null || !next.HasTrait<BaseBuilding>())
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs]: Why is this? From what I see, this change isn't needed. bases will contain all buildings if no conyards and only conyards when conyards, since you reused the variable above. This check was fine as-is.

File: OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs [hunk 7]
```diff
@@ -191,15 +191,24 @@ bool CycleBases()
...
+            // If no base is found all existing buildings are selected.
+            if (!bases.Any())
+            {
+				bases = world.ActorsWithTrait<TargetableBuilding>()
+                .Where(a => a.Actor.Owner == world.LocalPlayer)
+                .Select(b => b.Actor)
+                .ToList();
+                if (!bases.Any())
+					return true;
+            }
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs]: > OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs:L203: [SP2000] Invalid spacing at the end of the line.

You should run make check to notice issues like having additional whitespace at the end of a line.

Comment (penev92) [inline OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs]: Just in case you didn't understand that - OpenRA's Utility has a `make check` function, that checks the code for style errors and warns you.
It is what our CI system (Travis) is using, and this is why it's marked the build as failing below.

File: OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs [hunk 8]
```diff
@@ -191,15 +191,24 @@ bool CycleBases()
...
 
-			if (!bases.Any())
-				return true;
+            // If no base is found all existing buildings are selected.
+            if (!bases.Any())
+            {
+				bases = world.ActorsWithTrait<TargetableBuilding>()
+                .Where(a => a.Actor.Owner == world.LocalPlayer)
+                .Select(b => b.Actor)
+                .ToList();
```
Comment (WolfGaming) [inline OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs]: You seem to be using spaces here instead of tabs.

**Comment by pchote** [comment]
@bryanwilbur please join us in our IRC channel (http://www.openra.net/community/ - "Getting in Touch") if you'd like some help sorting out the indentation and git issues so that we can merge this.

**Comment by bryanwilbur** [comment]
I used make check and the tab issue and the extra spaces should be resolved.

File: OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs [hunk 9]
```diff
@@ -191,15 +191,24 @@ bool CycleBases()
...
 				.ToList();
 
+			// If no base is found all existing buildings are selected.
 			if (!bases.Any())
-				return true;
+			{
+				bases = world.ActorsWithTrait<TargetableBuilding>()
+				.Where(a => a.Actor.Owner == world.LocalPlayer)
+				.Select(b => b.Actor)
+				.ToList();
```
Comment (penev92) [inline OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs]: Please add one more level of indentation on the last 3 lines, and then an empty line after this.

File: OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs [hunk 10]
```diff
@@ -191,15 +191,24 @@ bool CycleBases()
...
+			}
 
 			var next = bases
 				.SkipWhile(b => !world.Selection.Actors.Contains(b))
 				.Skip(1)
 				.FirstOrDefault();
 
-			if (next == null)
+			// If no construction yard exists then next is set to only the oldest existing building
+			if (next == null || !next.HasTrait<BaseBuilding>())
```
Comment (penev92) [inline OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs]: What is the purpose of this additional check for `BaseBuilding`?

**Comment by penev92** [comment]
Forks :+1:

File: OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs (hunk 10 above)
Comment (bryanwilbur) [inline OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs]: Without that check using the hotkey will cycle through all buildings.  The check results in going to the oldest existing building which is more inline with how Red Alert handles returning to base without a construction yard.

---- On Sat, 04 Jul 2015 15:01:39 -0600 Pavel Penev &lt;notifications@github.com&gt; wrote ---- 

In OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs:
 &gt; &gt; var next = bases &gt; .SkipWhile(b =&gt; !world.Selection.Actors.Contains(b)) &gt; .Skip(1) &gt; .FirstOrDefault(); &gt; &gt; - if (next == null) &gt; + // If no construction yard exists then next is set to only the oldest existing building &gt; + if (next == null || !next.HasTrait&lt;BaseBuilding&gt;())  What is the purpose of this additional check for BaseBuilding?
 —
Reply to this email directly or view it on GitHub.
    
 
 




​

File: OpenRA.Mods.Common/Traits/Crates/Crate.cs [hunk 11]
```diff
@@ -76,10 +73,14 @@ public void OnCrush(Actor crusher)
 				var n = self.World.SharedRandom.Next(totalShares);
 
 				foreach (var s in shares)
-				{
+				{
 					if (n < s.Second)
 					{
+						if (!crusher.HasTrait<Armament>() && s.First.GetEffectName().Equals("firepower"))
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Crates/Crate.cs]: This doesn't appear to be related to the hotkey changes, and has its own issues that are best dealt with in its own pr.  Could you please split it off from this one?

**Comment by penev92** [comment]
Bump.
You should remove the changes to `Crate`, because those are not related to this PR, and address my style comment. Then we can merge this easily.

**Comment by bryanwilbur** [comment]
I am in the middle of moving, I will remove that commit next week.

Thanks,


Bryan

---- On Tue, 14 Jul 2015 13:28:07 -0600 Pavel Penev &lt;notifications@github.com&gt; wrote ---- 

Bump.
 You should remove the changes to Crate, because those are not related to this PR, and address my style comment. Then we can merge this easily.
 —
Reply to this email directly or view it on GitHub.
    
 
 





​

**Comment by pchote** [comment]
Thanks for the patches.  I have filed #8777 so that we can get this feature merged in time for the next playtest.
